rust/processor/src/gap_detector.rs

```diff
@@ -0,0 +1,183 @@
+    NoGapDetected(),
+}
+
+impl GapDetector {
+    pub fn new(processor_name: String, db_pool: PgDbPool, starting_version: u64) -> Self {
+        Self {
+            processor_name,
+            db_pool,
+            starting_version,
+            seen_versions: AHashMap::new(),
+            maybe_prev_batch: None,
+            num_batches_processed_without_gap: 0,
+            num_batches_processed_with_gap: 0,
+        }
+    }
+
+    pub async fn process_versions(
+        &mut self,
+        result: ProcessedVersions,
+    ) -> anyhow::Result<GapDetectorResult> {
+        // Check for gaps
+        let gap_detected = self.detect_gap(result.clone());
+
+        if gap_detected {
+            self.seen_versions.insert(result.start_version, result);
+            self.num_batches_processed_with_gap += 1;
+            tracing::debug!("Gap detected");
+        } else {
+            // If no gap is detected, find the latest processed batch without gaps
+            self.update_prev_batch(result);
+            self.num_batches_processed_with_gap = 0;
+            tracing::debug!("No gap detected");
+        }
+
+        // If there's a gap detected, panic.
+        // A gap is detected if we've processed GAP_DETECTION_BATCH_COUNT batches without seeing the batch that comes after prev_batch
+        if self.num_batches_processed_with_gap >= GAP_DETECTION_BATCH_COUNT {
+            let gap_start_version = if let Some(prev_batch) = &self.maybe_prev_batch {
+                prev_batch.end_version + 1
+            } else {
+                self.starting_version
+            };
+            return Ok(GapDetectorResult::GapDetected { gap_start_version });
+        }
+
+        // Check if need to update processor status
+        if self.num_batches_processed_without_gap >= PROCESSOR_STATUS_UPDATE_BATCH_COUNT {
+            let prev_batch = self.maybe_prev_batch.clone().unwrap();
+            self.update_processor_status(
+                prev_batch.end_version,
+                prev_batch.last_transaction_timstamp,
+            )
+            .await;
+            self.num_batches_processed_without_gap = 0;
+        }
+
+        Ok(GapDetectorResult::NoGapDetected())
+    }
+
+    fn detect_gap(&self, result: ProcessedVersions) -> bool {
+        if let Some(prev_batch) = &self.maybe_prev_batch {
+            prev_batch.end_version + 1 != result.start_version
+        } else {
+            result.start_version != self.starting_version
+        }
+    }
+
+    fn update_prev_batch(&mut self, result: ProcessedVersions) {
+        let mut new_prev_batch = result;
+        self.num_batches_processed_without_gap += 1;
+        while let Some(next_version) = self.seen_versions.remove(&(new_prev_batch.end_version + 1))
+        {
+            new_prev_batch = next_version;
+            self.num_batches_processed_without_gap += 1;
+        }
+        self.maybe_prev_batch = Some(new_prev_batch);
+    }
+
+    async fn update_processor_status(
```

Comment:
Isn't this copy pasting a bunch of code from existing files? Why not just reuse or at least clean up the old stuff.

Reply:
i'll clean up the old stuff